core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -51,6 +51,7 @@ import scala.math._
  */
 abstract class AbstractFetcherThread(name: String,
                                      clientId: String,
+                                     val leader: LeaderEndPoint,
```

Comment:
It's a bit weird to pass in both LeaderEndPoint and BrokerEndPoint since the former includes the latter. We could address that in a separate PR if easier.

Reply:
iiuc, the BrokerEndPoint isn't passed into the LeaderEndPoint (not Remote or Local)?

Reply:
Well, RemoteLeaderEndPoint takes a ReplicaFetcherBlockingSend, which takes a BrokerEndPoint, right?

Reply:
However, I don't think LocalLeaderEndPoint has any params with access to the brokerEndPoint, so we would need to pass it in as a param to LocalLeaderEndPoint

Reply:
Yes, we need to pass brokerEndPoint to LocalLeaderEndPoint. Intuitively, it seems that it makes sense for LeaderEndPoint to own brokerEndPoint instead of passing in two endpoints around.

Reply:
Yes, that does make sense. Probably it would be best to pass in BrokerEndPoint to LeaderEndpoint and construct a ReplicaFetcherBlockingSend within the RemoteLeaderEndPoint. However, ReplicaFetcherBlockingSend requires several other parameters (like metrics, time, fetcherId, etc.), which may make the constructor for RemoteLeaderEndPoint look a bit cluttered.

Reply:
We could. Alternatively, if we add brokerEndpoint() to LeaderEndPoint, we could implement it in RemoteLeaderEndPoint by getting it from the ReplicaFetcherBlockingSend inside RemoteLeaderEndPoint.

Reply:
We could end up having something like the following: 

```
trait LeaderEndPoint {
      def brokerEndPoint(): BrokerEndPoint
      ....
}

class LocalLeaderEndPoint(sourceBroker: brokerEndPoint, ....) extends LeaderEndPoint {
    def brokerEndPoint(): return sourceBroker

class RemoteLeaderEndPoint(endpoint: BlockingSend, ...) extends LeaderEndPoint {
    def brokerEndPoint(): return endpoint.sourceBroker
}
```

Reply:
Yes, then we just need to pass in a single endpoint to AbstractFetcherThread. This seems simpler, right?